libbpf-tools/funcinterval.c

```diff
@@ -0,0 +1,391 @@
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show this help", 0 },
+	{}
+};
+
+static inline
+long argp_parse_long(int key, const char *arg, const struct argp_state *state)
+{
+	long temp;
+
+	if (!arg) {
+		fprintf(stderr, "Arg is NULL\n");
+		argp_usage(state);
+	}
+
+	errno = 0;
+	temp = strtol(arg, NULL, 10);
+	if (errno || temp < 0) {
+		fprintf(stderr, "Error arg: %c : %s\n", (char)key, arg);
+		argp_usage(state);
+	}
+
+	return temp;
+}
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'p':
+		env.pid = argp_parse_long(key, arg, state);
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'i':
+		env.interval = argp_parse_long(key, arg, state);
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		if (state->arg_num != 0) {
+			fprintf(stderr, "Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		env.functions = arg;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format,
+			   va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = 1;
+}
+
+static inline
+const char *strftime_now(char *s, size_t max, const char *format)
+{
+	struct tm *tm;
+	time_t t;
+
+	t = time(NULL);
+	tm = localtime(&t);
+	if (!tm) {
+		fprintf(stderr, "localtime: %s\n", strerror(errno));
+		return "<failed>";
+	}
+	if (!strftime(s, max, format, tm)) {
+		fprintf(stderr, "strftime error\n");
+		return "<failed>";
+	}
+
+	return s;
+}
+
+static int print_log2_hists(struct funcinterval_bpf *obj)
+{
+	const char *units = env.milliseconds ? "msecs" : "usecs";
+	struct bpf_map *hists = obj->maps.hists;
+	int err, fd = bpf_map__fd(hists);
+	__u32 lookup_key = -1, next_key;
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	lookup_key = -1;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	fd = bpf_map__fd(obj->maps.start);
+	lookup_key = -1;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup start : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+static int split_pattern(const char *raw_pattern, enum TRACE_TYPE *type,
```

Comment:
The function name should be descriptive enough to convey its purpose without needing to read the code.